README
NAME
    CGI::Pure::Fast - Fast Common Gateway Interface Class for CGI::Pure.

SYNOPSIS
     use CGI::Pure::Fast;

     my $cgi = CGI::Pure::Fast->new(%parameters);
     $cgi->append_param('par', 'value');
     my @par_value = $cgi->param('par');
     $cgi->delete_param('par');
     $cgi->delete_all_params;
     my $query_string = $cgi->query_string;
     $cgi->upload('filename', '~/filename');
     my $mime = $cgi->upload_info('filename', 'mime');
     my $query_data = $cgi->query_data;

METHODS
    "new(%parameters)"
             Constructor.
             Extends CGI::Pure for FCGI.

     Other methods are same as CGI::Pure.

EXAMPLE
     use strict;
     use warnings;

     use CGI::Pure::Fast;
     use HTTP::Headers;

     # HTTP header.
     my $header = HTTP::Headers->new;
     $header->header('Content-Type' => 'text/html');

     # FCGI script.
     my $count = 1;
     while (my $cgi = CGI::Pure::Fast->new) {
             print $header->as_string."\n";
             print $count++."\n";
     }

     # Output in CGI mode:
     # Content-Type: text/html
     # 
     # 1
     # ...
     # Content-Type: text/html
     # 
     # 1
     # ...

     # Output in FASTCGI mode:
     # Content-Type: text/html
     # 
     # 1
     # ...
     # Content-Type: text/html
     # 
     # 2
     # ...

DEPENDENCIES
    CGI::Pure, FCGI, Readonly.

SEE ALSO
    CGI::Pure
        Common Gateway Interface Class.

    CGI::Pure::Save
        Common Gateway Interface Class for loading/saving object in file.

REPOSITORY
    <https://github.com/michal-josef-spacek/CGI-Pure-Fast>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2011-2021 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.10



Changes
0.10

0.09 2021-02-21T11:51:09+01:00
 - Add REPOSITORY section to doc.
 - Fix bugtracker.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update copyright years.

0.08 2018-01-22T11:48:45+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.

0.07 2018-01-22T09:27:07+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Update Module::Install to 1.18 version.
 - Update author name.
 - Update copyright years.

0.06 2015-09-26T10:25:40+02:00
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Remove obsolete item from SEE ALSO section in doc.
 - Update Module::Install to 1.16 version.
 - Update copyright years.

0.05 2014-11-02T12:31:33+01:00
 - Add EXAMPLE section to doc.
 - Fix functionality with Apache2 mod_fcgid.
 - Remove 'init' parameter check.
 - Rewrite B<> to C<> in doc.
 - Simplify code.
 - Update Module::Install to 1.14 version.

0.04 2014-10-17T07:33:26+02:00
 - Rewrite to new FCGI API.

0.03 2014-10-16T23:28:42+02:00
 - Add missing Test::NoWarnings test dependency.

0.02 2014-10-13T13:10:47+02:00
 - Add LICENSE file.
 - Add minimal version of Perl.
 - Fix MANIFEST file.
 - Rename CHANGES to Changes.
 - Update Module::Install to 1.12 version.

0.01 2013-02-26
 - First version.


Fast.pm
package CGI::Pure::Fast;

use base qw(CGI::Pure);
use strict;
use warnings;

use FCGI;
use Readonly;

# Constants.
Readonly::Scalar my $FCGI_LISTEN_QUEUE_DEFAULT => 100;

our $VERSION = 0.10;

# External request.
our $EXT_REQUEST;

# Workaround for known bug in libfcgi.
while (each %ENV) { }

# Constructor.
sub new {
	my ($class, @params) = @_;
	if (! defined $EXT_REQUEST) {
		if ($ENV{'FCGI_SOCKET_PATH'}) {
			my $path = $ENV{'FCGI_SOCKET_PATH'};
			my $backlog = $ENV{'FCGI_LISTEN_QUEUE'}
				|| $FCGI_LISTEN_QUEUE_DEFAULT;
			my $socket  = FCGI::OpenSocket($path, $backlog);
			$EXT_REQUEST = FCGI::Request(\*STDIN, \*STDOUT,
				\*STDERR, \%ENV, $socket, 1);
		} else {
			$EXT_REQUEST = FCGI::Request;
		}
	}
	if ($EXT_REQUEST->Accept < 0) {
		return;
	}
	return $class->SUPER::new(@params);
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

CGI::Pure::Fast - Fast Common Gateway Interface Class for CGI::Pure.

=head1 SYNOPSIS

 use CGI::Pure::Fast;

 my $cgi = CGI::Pure::Fast->new(%parameters);
 $cgi->append_param('par', 'value');
 my @par_value = $cgi->param('par');
 $cgi->delete_param('par');
 $cgi->delete_all_params;
 my $query_string = $cgi->query_string;
 $cgi->upload('filename', '~/filename');
 my $mime = $cgi->upload_info('filename', 'mime');
 my $query_data = $cgi->query_data;

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor.
 Extends CGI::Pure for FCGI.

=back

 Other methods are same as CGI::Pure.

=head1 EXAMPLE

 use strict;
 use warnings;

 use CGI::Pure::Fast;
 use HTTP::Headers;

 # HTTP header.
 my $header = HTTP::Headers->new;
 $header->header('Content-Type' => 'text/html');

 # FCGI script.
 my $count = 1;
 while (my $cgi = CGI::Pure::Fast->new) {
         print $header->as_string."\n";
         print $count++."\n";
 }

 # Output in CGI mode:
 # Content-Type: text/html
 # 
 # 1
 # ...
 # Content-Type: text/html
 # 
 # 1
 # ...

 # Output in FASTCGI mode:
 # Content-Type: text/html
 # 
 # 1
 # ...
 # Content-Type: text/html
 # 
 # 2
 # ...

=head1 DEPENDENCIES

L<CGI::Pure>,
L<FCGI>,
L<Readonly>.

=head1 SEE ALSO

=over

=item L<CGI::Pure>

Common Gateway Interface Class.

=item L<CGI::Pure::Save>

Common Gateway Interface Class for loading/saving object in file.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/CGI-Pure-Fast>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2011-2021 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.10

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Fast Common Gateway Interface Class for CGI::Pure.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.01;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'CGI-Pure-Fast';
readme_from 'Fast.pm';
recursive_author_tests('xt');
requires 'CGI::Pure' => 0.01;
requires 'FCGI' => 0.77;
requires 'Readonly' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/CGI-Pure-Fast/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/CGI-Pure-Fast';
resources 'repository' => 'git://github.com/michal-josef-spacek/CGI-Pure-Fast';
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.10';

# Run.
WriteAll();
